# Ejercicio 1

Implementa el algoritmo extendido de Euclides para el cálculo del máximo común divisor: dados dos enteros $a$ y $b$, encuentra $u, v ∈ Z$ tales que $au + bv$ es el máximo común divisor de $a$ y $b$.  

Para la realización del algoritmo, nos basamos en la referencia aportada en el _pdf_, en concreto en el [Algoritmo 2.107](http://cacr.uwaterloo.ca/hac/about/chap2.pdf).

### Pseudocódigo del Algoritmo Extendido de Euclides

**ENTRADA**: Dos enteros positivos $a$ y $b$ donde $a ≥ b$  
**SALIDA**: $d = mcd(a, b)$ y enteros $u$, $v$ donde $a(u) + b(v) = d$     
  
- Si $b = 0$ hemos terminado el algoritmo y se devuelve $d = a$, $u = 1$, $v = 0$
- Sino $u_0 = 1$, $u_1 = 0$, $v_0 = 0$, $v_1 = 1$
    - Mientras $b > 0$ hacer
        - $q = a\ div\ b$ 
        - $(a, b) = (b, a - q·b)$
        - $(u_0, u_1) = (u_1, u_0 - q·u_1)$
        - $(v_0, v_1) = (v_1, v_0 - q·v_1)$  
        
    - Se devuelve $d = a$, $u = u_0$, $v = v_0$
   
_**Nota**: Si el valor es negativo, le cambiamos el signo._

In [1]:
# -*- coding: utf-8 -*-

# Ahora procedemos a calcular el Algoritmo Extendido de Euclides
# donde a y b son los enteros positivos que se pasan como entrada
def alg_euclides(a, b):
    
    # Nos creamos una lista que devolverá los valores de la ecuación
    euclides = []
    
    # Si el valor es negativo de ambos valores
    # los convertimos a positivos
    if a < 0 or b < 0:
        a = -(a) # Convierto 'a' en positivo
        b = -(b) # Convierto 'b' en positivo
    
    # Si a = 0 y b = 0 hemos terminado y se devuelve "Error"
    if a == 0 and b == 0:
        print("Error, los dos números introducidos son cero.")
    
    # Si a = 0 hemos terminado y se devuelve 0
    if a == 0:
        d = 0
        u = 0
        v = 0
    
    # Si b = 0 hemos terminado y se devuelve d = a, u = 1, v = 0
    if b == 0:
        d = a
        u = 1
        v = 0
    
    # En caso contrario donde a y b sean positivos
    else:
        
        # Comprobamos si a < b, en caso afirmativo
        # cambiamos el orden de los valores dejando a > b
        if a < b:
            tmp = a
            a = b
            b = tmp
    
        # Guardamos el valor de la entrada, para no perderlos
        a0 = a
        b0 = b
    
        # Inicializamos las variables a manejar
        u0 = 1
        u1 = 0
        v0 = 0
        v1 = 1
        
        # Mientras b > 0
        while b > 0:
            
            # q = a div b
            q = a // b
            
            # (a, b) = (b, a-q*b)
            r = a - q*b
            a = b
            b = r
            
            # (u0, u1) = (u1, u0-q*u1)
            u = u0 - q*u1
            u0 = u1
            u1 = u
            
            # (v0, v1) = (v1, v0-q*v1)
            v = v0 - q*v1
            v0 = v1
            v1 = v
         
        # Guardamos el valor que nos interesa
        d = a
        u = u0
        v = v0
        
    # Añadimos a la lista los valores para la ecuación
    euclides.append(a0)
    euclides.append(u)
    euclides.append(b0)
    euclides.append(v)
    euclides.append(d)

    # Devolvemos la lista
    return euclides

In [2]:
alg_euclides(267, 393)

[393, -36, 267, 53, 3]

### Interpretación del Resultado Obtenido    

Nos devuelve una lista como: $[a_0, u, b_0, v, d]$ donde $a_0(u) + b_0(v) = d$ siendo $a_0$ y $b_0$ los valores introducidos por parámetro. 

En este ejemplo **alg_euclides(393, 267)**, la ecuación quedaría de la siguiente manera: $393(-36) + 267(53) = 3$ donde $u = -36$ y $v = 53$

Veamos como se ha obtenido el resultado:

$q$ | $r$ | $u$ | $v$ | $a$ | $b$ | $u_0$ | $u_1$ | $v_0$ | $v_1$
:---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---:  
 $-$ | $-$ | $-$ | $-$ | * $393$ | * $267$  | $1$ | $0$ | $0$ | $1$
 $1$ | $126$ | $1$ | $-1$ | $267$ | $126$ | $0$ | $1$ | $1$ | $-1$
 $2$ | $15$ | $-2$ | $3$ | $126$ | $15$ | $1$ | $-2$ | $-1$ | $3$
 $8$ | $6$ | $17$ | $-25$ | $15$ | $6$ | $-2$ | $17$ | $3$ | $-25$
 $2$ | $3$ | $-36$ | $53$ | $6$ | $3$ | $17$ | $-36$ | $-25$ | $53$
 $2$ | $0$ | $-$ | $-$ | * $3$ | $0$ | * $-36$ | $-$ | * $53$ | $-$

In [3]:
# Como lo que nos interesa es usar números muy grandes, vamos 
# a probar con otros números, para ver que el AEC, lo calcula bien
alg_euclides(28564333765949, 123456789101119)

[123456789101119, -2517705255944, 28564333765949, 10881675356013, 1]

In [4]:
# Comprobamos que el resultado de antes es correcto
((123456789101119*-2517705255944) + (28564333765949*10881675356013))

1

Para comprobar que el resultado es correcto, podemos usar la función implementada ya en **GAP**: 
> GcdRepresentation(28564333765949, 123456789101119);

---

# Ejercicio 2 

Usando el ejercicio anterior, escribe una función que calcule $a^{-1} \pmod b$ para cualesquiera $a, b$ enteros que sean primos relativos.

Dos números son **primos relativos** si los únicos divisores que tienen en común, son el $1$ y el $-1$. Para ello haremos uso del **Teorema de Bezout** que dice que dos números enteros $a$ y $b$ son primos relativos, si y solo si, existen números enteros $u$ y $v$, tal que $a(u) + b(v) = 1$. Por tanto, para que un número tenga inverso, el máximo común divisor de ambos debe ser $1$.

Para realizar este ejercicio, nos basamos en el **pdf de Jesús Miranda de Criptografía**, en concreto en la _página 175_ o _185_, abierto desde el visualizador de pdf's. 

### Pseudocódigo

1. Calculamos el Algoritmo Extendido de Euclides para $a$ y $b$
2. Si $mcd(a,b)\ != 1$  
    2.1\. No existe el inverso
3. Si $mcd(a,b) = 1$  
    3.1\. Existe el inverso, el cual será la variable $v$ del Algoritmo Extendido de Euclides. Donde $a*u + b*v = 1$, siendo $v$ el inverso de $a$  
    3.2\. Si el inverso es negativo, lo convertiremos en postivo con $\pmod {a}$ 

In [5]:
# -*- coding: utf-8 -*-

# Algoritmo que calcula el inverso de dos numeros 
# que son primos relativos
def alg_inverso(a, b):
    
    # Calculamos el Algoritmo Extendido de Euclides para a y b
    euclides = alg_euclides(a, b)
        
    # Si el mcd es distinto de 1, no tiene inverso    
    if euclides[4] != 1:
        print("\nNo existe inverso para {0}^(-1) en Z({1})".format(a,b))
        
    # Si el mcd es igual a 1, si tiene inverso  
    if euclides[4] == 1:
    
        # a(u) * b(v) = d
        # Como [a*u (mod a)] = 0. Nos quedamos con b*v
        # siendo el v el inverso de a
        
        # Si el inverso es positivo, es decir, v, lo dejo tal cual
        if (euclides[3] > 0) :
            return euclides[3] % b
        
        # Si el inverso es negativo, le cambio el signo en módulo 'a'
        elif (euclides[3] < 0):
            
            nuevo = b + euclides[3]
            
            # Mientras el valor siga siendo negativo, le vamos sumando 'a'
            while nuevo < 0:
                nuevo = b + euclides[3]
                
            return nuevo % b

In [6]:
alg_inverso(391,1542)

631

In [7]:
alg_euclides(391,1542)

[1542, -160, 391, 631, 1]

### Interpretación del Resultado Obtenido    

Nos devuelve la solución del _Algoritmo Extendido de Euclides_: $[a, u, b, v, d]$ donde $a*u + b*v = d$ siendo $a$ y $b$ los valores introducidos por parámetro. 

En este ejemplo **alg_inverso(391, 1542)**, la ecuación quedaría de la siguiente manera: $1542(-160) + 391(631) = 1$ donde $u = -160$ y $v = 631$ como $1542*-160 \pmod 1542 = 0$, nos quedamos con $v$

Veamos como se ha obtenido el inverso:

$(a, b) = (1542,391)_{q=3} = (391,369)_{q=1} = (369,22)_{q=16} = (22,17)_{q=1} = (17,5)_{q=3} = (5,2)_{q=2} = (2,1)_{q=2} = (1,0)$  
  
$(u_0, u_1) = (1,0)_{q=3} = (0,1)_{q=1} = (1,-1)_{q=16} = (-1,17)_{q=1} = (17,-18)_{q=3} = (-18,71)_{q=2} = (71,-160)_{q=2} = (-160,\_)$  
  
$(v_0, v_1) = (0,1)_{q=3} = (1,-3)_{q=1} = (-3,4)_{q=16} = (4,-67)_{q=1} = (-67,71)_{q=3} = (71,-280)_{q=2} = (-280,631)_{q=2} = (631,\_)$  

Lo que acabamos de hacer, es calcular el _Algoritmo Extendido de Euclides_, no hacía falta calcular todo, ya que con $(v_0, v_1)$, hubiéramos llegado al mismo resultado.

Por tanto, el inverso de $391^{-1} \pmod {1542} = 631$

In [8]:
# Como lo que nos interesa es usar números muy grandes, vamos 
# a probar con otros números, para ver que el inverso, lo calcula bien.
alg_inverso(28564333765949, 123456789101119)

10881675356013

In [9]:
# Comprobamos que el mcd de ambos números es 1
(28564333765949*_)%123456789101119

1

Para comprobar que el resultado es correcto, podemos usar la función implementada ya en **GAP**: 
> GcdRepresentation(28564333765949, 123456789101119)[1];

---

# Ejercicio 3 

Escribe una función que calcule $a^k \pmod n$ para cualesquiera $a$, $k$ y $n$ enteros positivos. La implementación debería tener en cuenta la representación binaria de k. _(No hace falta calcular la expresión binaria)_

Para realizar este ejercicio, nos basamos en la referencia aportada en el ejercicio, en concreto en el [Algoritmo 2.143](http://cacr.uwaterloo.ca/hac/about/chap2.pdf)

### Pseudocódigo

**ENTRADA**: $a\in \mathbb{Z}_n$, y dos enteros $0 <= k < n$  
**SALIDA**: $a^k \pmod n$  

1. Inicializar $b = 1$  
**{$k = {k}_0 · {k}_1 · .... · {k}_r$}**
2. Mientras $k > 0$  
    2.1\. ${k}_0 = k \pmod 2$  
    2.2\. Si ${k}_0 = 0$ entonces $b = a·b \pmod n$  
    2.3\. $k = (k - {k}_0) \div 2$  
    2.4\. $a = a^2 \pmod n$
3. Devolvemos $b$

In [10]:
# -*- coding: utf-8 -*-

# Algoritmo que calcula el a^k mod n
def alg_potencia(a, k, n):
    
    # Inicializamos b a 1
    b = 1

    # Si k = 0, ya hemos terminado y devolvemos 1
    
    # Pero mientras k > 0    
    while k > 0:
        
        # Iremos descomponiendo k0 
        # Y comprobaremos si su resto es 0 ó 1
        k0 = k % 2
        
        # Si el resto es 1
        if k0 == 1:
            
            # Cambiamos el valor de b
            b = a*b % n
        
        # Como k = k0 ... kr
        # Nos quedamos con el k sin usar
        k = (k-k0) / 2 
        
        # Cambiamos el valor de a
        a = (a**2) % n

    # Devolvemos b    
    return b

In [11]:
alg_potencia(5, 596, 1234)

1013

### Interpretación del Resultado Obtenido    

Al hacer **alg_potencia(5, 596, 1234)**, lo que estamos haciendo es $5^{596} \pmod {1234}$, el resultado de todo esto es $1013$. ¿Pero cómo hemos obtenido ese resultado? Veámoslo gráficamente:

_Recuerda que estamos en $\pmod {1234}$_

$iteracion$ | ${k}_i$ | $Valor\ de\ a$ | $Valor\ de\ b\ =(a*b)$
:---: | :---: | :---: | :---:  
0 | 0 | 5² = 25 | 1 
1 | 0 | 25² = 625 | 1 
2 | 1 | 625² = 681 | 625·1 = 625 
3 | 0 | 681² = 1011 | 625·1 = 625
4 | 1 | 1011² = 369 | 1011·625 = 67 
5 | 0 | 369² = 421 | 1011·625 = 67
6 | 1 | 421² = 779 | 421·67 = 1059 
7 | 0 | 779² = 947 | 421·67 = 1059 
8 | 0 | 947² = 925 | 421·67 = 1059 
9 | 1 | 925² = 463 | 925·1059 = 1013 

In [12]:
# Como lo que nos interesa es usar números muy grandes, vamos 
# a probar con otros números, para ver que la potencia, lo calcula bien.
alg_potencia(8590365927553, 28564333765949, 123456789101119)

65615831162116

Para comprobar que el resultado es correcto, lo comprobamos haciendo en **GAP**:
> PowerModInt(8590365927553, 28564333765949, 123456789101119);

---

# Ejercicio 4

Dado un entero $p$, escribe una función para determinar si $p$ es (probablemente) primo usando el **método de Miller-Rabin**.

Para realizar este ejercicio, nos basamos en el **pdf de Jesús Miranda de Criptografía**, en concreto en la _página 193_ o _199_, abierto desde el visualizador de pdf's. 

### Pseudocódigo

**ENTRADA**: Un número entero que sea $p >= 5$  
**SALIDA**: Si $p$ es problamente primo o no.

1. Descompongo $p-1$  
    1.1\. $p-1 = 2^{u} * s$  
    1.2\. Elijo aleatoriamente un número de $a \in{}_R \{2, ..., p-2\}$
2. Calculo $a^{s}$  
    2.1\. Si es $1$ o $-1$ devuelvo _probablemente primo_  
    2.2\. Si no es $1$ o $-1$ calculo $a^{2*s}$ y luego voy elevando al cuadrado como $(a^{2*s})^{2} = a^{2^{2}*s} = ... = a^{2^{u}*s}$ 

In [13]:
# -*- coding: utf-8 -*-

# Algoritmo que calcula la primalidad de un numero
import random

def alg_miller_rabin(p):
    
    # Primero sacamos la primalidad para los números del 1 al 5
    
    # Si el número es 1 o 4, no es primo
    if p == 1 or p == 4:
        return False # No primo
    
    # Si el número es 2, 3, es primo
    if p == 2 or p == 3:
        return True #Primo
        
    # En caso de que el número sea mayor que 5
    # Descompongo en p-1
    p1 = p - 1
    u = 0

    # Descompongo p-1 en 2^(u) * s
    while p1%2 == 0:
        p1 = p1 // 2
        u = u + 1
    
    s = p1

    # Eligo aleatoriamente un número entre 2 y p-2
    a = random.randint(2, p-2)

    # Calculo a^(s) -> a = a**s
    a = alg_potencia(a,s,p)

    # Y compruebo si es 1 o-1,
    if a == 1 or a == p-1:
        return True # Problamente primo

    # Sino es primo, voy elevando al cuadrado    
    for i in range(1, u):
        a = alg_potencia(a, 2, p) # a = (a**2) % p

        # Si 'a' es p-1
        if a == p-1:
            return True # Problamente primo

        # Si 'a' es 1
        if a == 1:
            return False # Probalmemente no primo

    return False # Probalmemente no primo

In [14]:
alg_miller_rabin(21)

False

### Interpretación del Resultado Obtenido    

Ejecutamos **alg_miller_rabin(21)**, siendo nuestro número $p = 1$.  
  
Por tanto $p-1 = 20 = 2^{2} * 5$ siendo $u = 2$ y $s = 5$. Escojo aleatoriamente un número de la siguiente lista: $a \in{}_R \{2, ..., 19\}$  
- Si $a = 7$ -> $7^5 = 7$, $7² = 7$, $7² = 7$ _(voy elevando al cuadrado a partir del segundo)_
- Si $a = 4$ -> $4^5 = 16$, $16² = 4^{2*5} = 4$, $4^{2^{2}*5} = 16$ _(voy elevando al cuadrado a partir del segundo)_

Como no aparece ningún $1$ o $-1$, $21$ es probablemente no primo.

La función que hemos creado, solo se ejecuta 1 vez, ahora hay que crear otra función que se llame $n$ veces, para comprobar si ese número es primo o no.

In [15]:
def alg_miller_rabin1(p, k):
    return all(alg_miller_rabin(p) for a in range(k))

In [16]:
alg_miller_rabin1(21,10)

False

A continuación, vamos a calcular el test de Miller-Rabin para la lista de primos proporcionados por la práctica en el pdf.

In [17]:
# Lista de números primos
lista=[46381, 768479, 9476407, 36780481, 562390847, 1894083629, 65398261921, 
       364879542899, 8590365927553, 28564333765949, 123456789101119]

for i in lista:
    print("El numero {0} es {1}".format(i, "primo" if alg_miller_rabin1(i,10) else "no primo"))

El numero 46381 es primo
El numero 768479 es primo
El numero 9476407 es primo
El numero 36780481 es primo
El numero 562390847 es primo
El numero 1894083629 es primo
El numero 65398261921 es primo
El numero 364879542899 es primo
El numero 8590365927553 es primo
El numero 28564333765949 es primo
El numero 123456789101119 es primo


Para comprobar que el resultado es correcto, lo comprobamos haciendo en **GAP**:
> IsPrime(123456789101119);

----

# Ejercicio 5

Implementa el algoritmo **paso enano-paso gigante** para el cálculo de algoritmos discretos en $\mathbb{Z}_p$.

Para la realización de este ejercicio, nos hemos basado en los apuntes copiados en clase (_página 15 de mis apuntes_).

### Pseudocódigo

**ENTRADA**: $a\in \mathbb{Z^*}_p$, $b\in \mathbb{Z^*}_p$ y $p$ siendo primo.  
**SALIDA**: ¿Existe un $k$ con $a^k = b$?  

1. Encontrar un $s$ tal que $s^2 ≥ p$ y $s = \sqrt{(p-1)}$, donde escogemos la parte entera redondeada hacia arriba.
2. Calcular los valores que tendrá la lista para el paso gigante. La lista empieza en el índice 1 hasta s+1  
    2.1\. $L = \{a^{s}, a^{2·s}, ..., a^{s·s}\}$ todo en $\pmod{p}$
3. Calcular los valores que tendrá la lista para el paso enano. La lista empieza en el índice 0 hasta s-1  
    3.1\. $l = \{b·a^{0}, b·a^{1}, ..., b·a^{s-1}\}$ todo en $\pmod{p}$
4. Recorremos las dos listas, buscando el elemento común y la posición de ese elemento en ambas listas.
5. Devolvemos el $k$ que será $a^k = b$, siendo $k = \{{L[posicion+1]}_{elemento-comun} * s - {l[posicion]}_{elemento-comun} \}\pmod p$  

In [18]:
# -*- coding: utf-8 -*-

# Función para calcular la raíz cuadrada de un número
def isqrt(n):
    x = n
    y = (x+1) // 2
    
    while(y < x):
        x = y
        y = (x+n // x) // 2
        
    return x

# Función que calcula el Algoritmo de
# Paso Enano Paso Gigante
def paso_enano_paso_gigante(a, b, p):
    
    # Nos creamos una lista con todos los posibles a^(cs)
    # Esta lista será el paso gigante
    L = []
    
    # Nos creamos una lista con todos los posibles b*a^(r)
    # Esta lista será el paso enano
    l = []
    
    # Obtenemos s = [sqrt(p-1)], con redondeo hacia arriba
    s = isqrt(p-1)+1
    
    # Calculamos los valores que tendrá la lista para el
    # paso gigante que empieza el índice en 1 hasta s+1
    for i in range(1, s+1):
        # Los valores de la lista serán [a^(i*s) % p]
        gigante = alg_potencia(a, i*s, p)
        L.append(gigante)
    
    # Calculamos los valores que tendrá la lista para el
    # paso enano que empieza el índice en 0 hasta s-1
    for i in range(0, s):
        # Los valores de la lista serán [b*a^(i) % p]
        enano = (b*alg_potencia(a, i, p))% p;
        l.append(enano)
        
    # Nos creamos una variable, donde se guardará la
    # posición de ambas listas
    aux_l = 0.0
    aux_L = 0.0
    
    # Recorremos la primera lista y buscamos ese elemento
    # en la siguiente lista
    for x in L:
        
        # Si tienen ambas listas un elemento común
        if x in l:
            
            # Posición donde se encuentra el elemento
            # repetido en la lista L
            aux_L = L.index(x)
            
            # Posición donde se encuentra el elemento
            # repetido en la lista l
            aux_l = l.index(x)
               
    # k = c*s - r = (aux_L+1)*s - (aux_l) = (aux_L+1)*s-aux_l) % p)
    return (((aux_L+1)*s-aux_l) % p)

In [19]:
paso_enano_paso_gigante(6, 32, 41)

10

### Interpretación del Resultado Obtenido    

Ejecutamos **paso_enano_paso_gigante(6, 32, 41)**, siendo $a = 6$, $b = 32$ y $p = 41$. Por tanto:  
- $s^2 >= p-1 = 40$ 
- $s = \sqrt{(p-1)} = \sqrt{40} = 6'32 = 7$ (_debemos escoger la parte entera redondeando hacia arriba_)  
  
Calculamos la lista del _paso gigante_, desde $1$ hasta $s$:  
- $L = \{6^{7}, 6^{2·7}, 6^{3·7}, 6^{4·7}, 6^{5·7}, 6^{6·7}, 6^{7·7}\}$ todo en $\pmod{41}$ sería $L = \{29, 21, 35, 31, 38, 36, 19\}$  
  
Calculamos la lista del _paso enano_, desde $0$ hasta $s-1$:  
- $l = \{32·6^{0}, 32·6^{1}, 32·6^{2}, 32·6^{3}, 32·6^{4}, 32·6^{5}, 32*6^{6}\}$ todo en $\pmod{41}$ sería $l = \{32, 28, 4, 24, 21, 3, 18\}$  
  
Obtenemos que el elemento que se repite es el $21$, y se encuentra en la posición:  
- En la posición $[2]$ de la lista $L$ (_empezamos a contar desde_ $1$)  
- En la posición $[4]$ de la lista $l$ (_empezamos a contar desde_ $0$)  
  
Por consiguiente, $6^{2·7} = 32·6^{4}$, luego $32 = 6^{2·7-4}$. Por tanto, $k = c*s - r = 2*7 - 4 = 10$

Para comprobar que el resultado es correcto, lo podemos comprobar usando la función implementada en **GAP**:
> LogModShanks(a,b,m);

---

# Ejercicio 6 

## Ejercicio 6.1

Sea $n = pq$, con $p$ y $q$ enteros primos positivos.

1. Escribe una función que, dado un entero $a$ y un primo $p$ con $(\frac{a}{p}) = 1$, devuelve $r$ tal que $r² = a \pmod p$; primero te hará falta implementar el símbolo de Jacobi.

### Jacobi

Para la realización del algoritmo de Jacobi, nos basamos en la referencia aportada en el _pdf_, en concreto en el [Algoritmo 2.149](http://cacr.uwaterloo.ca/hac/about/chap2.pdf).

#### Pseudocódigo

**ENTRADA**: Dos enteros, donde $p ≥ 3$ y donde $a$ sea $0 ≤ a < n$  
**SALIDA**: El simbolo de Jacobi (y por lo tanto, el símbolo de Legendre cuando $p$ es primo)

1. Si $a = 0$ se devuelve $0$
2. Si $a = 1$ se devuelve $1$
3. Ir dividiendo $a$ entre $2$ hasta que sea impar ($a = 2^ea_1$).  A la vez vamos calculando el número de veces que se ha dividido el número, para ello usamos la variable $e$
4. Si $e \div 2 = 1$ devuelvo $s = 1$  
    4.1\. Sino, si $p ≡ 1\ or\ 7 \pmod 8$ devuelvo $s = 1$  
    4.2\. Sino, si $p ≡ 3\ or\ 5 \pmod 8$ devuelvo $s = -1$  
5. Si $p ≡ 3 \pmod 4$ y $a ≡ 3 \pmod 4$ devuelvo $s = -s$  
6. Divido $p \div a$
7. Si $a = 1$ devuelvo $s$
8. Sino, devuelvo $(s * jacobi(p,a))$

In [20]:
# -*- coding: utf-8 -*-

# Función que calcula si un número es símbolo de Jacibo
def jacobi(a, p):
        
    # Si a es 0 se devuelve 0
    if a == 0:
        return 0
    
    # Si a es 1 se devuelve 1
    if a == 1:
        return 1
    
    # Variable que cuenta el numero de veces que se 
    # ha dividido el número
    e = 0
    
    # Vamos diviendiendo 'a' hasta que sea impar
    while a % 2 == 0:
        a /= 2
        e += 1
      
    # Si e % 2 = 1, devuelvo s=1
    if e % 2 == 0:
        s = 1 
    else:  
        # Si p ≡ 1 or 7 (mod 8) devuelvo s=1
        if p%8 == 1 or p%8 == 7:
            s = 1

        # Si p ≡ 3 or 5 (mod 8) devuelvo s=-1
        elif p%8==3 or p%8==5:
            s = -1
    # Si p ≡ 3 (mod 4) y a ≡ 3 (mod 4) devuelvo s=-s
    if (p%4) == 3 and (a%4) == 3: 
        s = -s
    
    # Divido 'p' entre 'a'
    p = p % a
    
    # Si a es 1, devuelvo s
    if a==1:
        return s
    # Sino deevuelvo (s*jacobi1(p,a))
    else:
        return (s*jacobi(p,a))

In [21]:
# Para comprobar que funciona, vamos a realizar un ejemplo
jacobi(5, 17)

-1

### Interpretación del Resultado Obtenido    

Partimos de la _página 22 y 23 de mis apuntes_:

Ejecutamos **jacobi(5,17)**, siendo $a = 5$, $p = 17$. Por tanto:  
$$\frac{a}{p} = \frac{5}{17} {(1)}  = \frac{17}{5} {(2)}  = \frac{2}{5} {(3)}  = (-1)^{\frac{25-1}{8}} = (-1)^{3} = -1$$

- (1): Le doy la vuelta, no cumple la Ley de Criprocidad Cuadrática (propiedad 6)
- (2): Aplicamos la propiedad 1, que dice que $\frac{a}{p} = \frac{a \pmod p}{p}$ y como $\frac{17 \pmod 5}{5} = \frac{2}{5}$
- (3): Aplicamos la propiedad 3, que dice que $\frac{2}{p} = (-1)^{\frac{p^{2}-1}{8}}$ y como $\frac{2}{5}  = (-1)^{\frac{5^{2}-1}{8}} = (-1)^{\frac{25-1}{8}} = (-1)^{\frac{24}{8}} = (-1)^{3}$


Para comprobar que el resultado es correcto, lo comprobamos haciendo en **GAP**:
> Jacobi(5,17);

### Residuo Cuadrático

Ahora, pasamos a calcular el residuo cuadrático. Para ello, nos basamos en el **pdf de Jesús Miranda de Criptografía**, en concreto en la _página 182_ o _188_, abierto desde el visualizador de pdf's. 

#### Pseudocódigo

**ENTRADA**: Supongamos que tenemos un número primo $p$, un número $a$ tal que $\frac{a}{p} = 1$.  
**SALIDA**: La raíz cuadrada de $a \pmod p$

- Si $jacobi(a,p) = 1$
- Buscamos $n$ tal que $\frac{n}{p} = -1$
- Descomponemos $p-1$ como $p-1 = 2^{u}·s$, con $s$ un número impar.
- Si $u = 1$ hacemos $r = a^{\frac{p+1}{4}}$
- Si $u >= 2$ entonces:
    - Hacemos $r = a^{\frac{s+1}{2}}$
    - Hacemos $b = n^s$
    - Hacemos $j = 0$ 
    - Mientras $j <= u-2$
        - Si $(a^{-1}·r^{2})^{2^{u-2-j}} = -1$ hacemos $r = r·b$ 
        - Hacemos $b = b^2$
        - Hacemos $j = j + 1$
- Devuelve $r$

In [22]:
# -*- coding: utf-8 -*-

# Algoritmo que calcula la primalidad de un numero
import random

def residuo_cuadratico(a, p):
    
    if jacobi(a,p) == 1:
        i = 1
        n = 0
        r = 0
        
        # Buscamos aleatorio entre entre 2 y p-1 
        # tal que (i/p) = -1
        while (i != -1):
            # el randint coge hasta p-1 por eso solo ponemos p
            n = random.randint(2,p)
            i = jacobi(n,p)
        
        # Descomponemos p-1 en 2^u * s
        p1 = p-1
        u = 0
        
        while p1%2 == 0:
            p1 = p1 // 2
            u = u + 1
    
        s = p1
        
        # Si 'u' es igual 1
        if u == 1:
            r = alg_potencia(a, (p+1)//4, p)
        
        # Sino 
        elif u >= 2:
            r = alg_potencia(a, (s+1)//2, p)
            b = alg_potencia(n, s, p)
            
            j = 0
            
            a_i = alg_inverso(a, p)
            
            # Mientras j <= u-2
            while (j <= u-2):
                if alg_potencia(a_i*r**2, 2**(u-2-j), p) == p-1:
                    r = (r*b) % p 
                    
                b = (b*b)%p
                j = j + 1
        
        # Devolvemos r
        return r                      
        
    # elif jacobi1(a,p) == 1: ->  No tiene raíces
    # Sino devuelve nada, no tiene raíces

In [23]:
residuo_cuadratico(319, 353)

242

Hay que tener en cuenta, que existen dos raíces $111$ y $242$, así de que dependiendo de la ejecución, nos podrá dar una u otra. Para comprobar que está bien, hacemos:

In [24]:
alg_potencia(_, 2, 353)

319

### Interpretación del Resultado Obtenido    

Partimos de la _página 184 o 190, abierto desde el visualizador de pdf's_ de los **apuntes de Miranda**.

Vamos a calcular la raíz cuadrada de $319 \pmod 353$. En primer lugar comprobamos si dicha raíz existe (bueno, habría que ver antes que 353 es primo).

$$ \frac{319}{353} = \frac{353}{319} = \frac{34}{319} = \frac{2}{319} · \frac{17}{319} = \frac{17}{319} = \frac{319}{17} = \frac{13}{17} = \frac{17}{13} = \frac{4}{13} = 1 $$

Una vez hecho esto, necesitamos expresar $352$ como $2^u·s$, y encontrar $n$ tal que $\frac{n}{353} = -1$  
Para lo primero vamos dividiendo $352$ por dos hasta que nos dé impar:

$$ 352 -> 176 -> 88 -> 44 -> 22 -> 11 $$

es decir, $352 = 2^5·11$ ($u = 5$; $s = 11$).  
Para encontrar $n$, vamos probando con $2, 3, ...$. hasta que obtengamos uno.

- $\frac{2}{353} = 1$ pues $353 = 1 \pmod 8$
- $\frac{3}{353} = \frac{325}{3} = \frac{2}{353} = -1$

Por tanto, $n = 3$ _(o en nuestro caso la $i$)_  
Inicializamos las contantes:  
$r = a^{\frac{s+1}{2}} = 319^6$  
$b = n^s = 3^{11} = 294$  
Y calculamos $a^{-1}$, que vale $218$.

$j$ | $a^{-1}r^{2}$ | $(a^{-1}$r²$)^{2^{3-j}}$ | $r$ | $b$
:--: | :--: | :--: | :--: |:--:
 |  | | 168 | 294 
0 | 42 | 1 |  | 304
1 | 42 | 1 |  | 283 
2 | 42 | -1 | 242 | 311
3 | 1 | 1 |  | 352 

El valor de $r$, es $242$ que es una raíz cuadrada de $319$. La otra raíz cuadrada es $353-242=111$. 

Para comprobar que el resultado es correcto, lo comprobamos haciendo en **GAP**:
> RootsMod(319,353);

## Ejercicio 6.2. 

Sea $n = pq$, con $p$ y $q$ enteros primos positivos.

2. Sea a un entero que es residuo cuadrático módulo $p$ y $q$. Usa el teorema chino de los restos para calcular todas las raíces cuadradas de $a \pmod n$ a partir de las raíces cuadradas de $a$ módulo $p$ y $q$.

### Teorema Chino de los Restos

Para calcular el teorema chino de los restos, me baso en mis apuntes tomandos en clase (_página 18_). En concreto, en la fórmula:  

$$x = rp + p·(rq-rp)·p^{-1}$$

In [25]:
# -*- coding: utf-8 -*-

# Algoritmo que calcula el teorema chino de los restos
def teorema_chino_restos(rp, rq, p, q):
    n = p*q
    inverso_p = alg_inverso(p, q)
    return ((rp + p *(rq - rp)*inverso_p) % (n))

### Calcular Raíces

Para calcular las raíces, me baso en mis apuntes tomandos en clase (_página 19_).

In [26]:
# -*- coding: utf-8 -*-

# Algoritmo que calcula las raíces modulares
def raices(a, p, q):
    n = p * q
    
    # Calculo el residuo cuadrático de p y q
    rp = residuo_cuadratico(a,p)
    rq = residuo_cuadratico(a,q)
    
    # jacobi != 1, no es residuo cuadratico
    if (rp != None and rq != None):
        # A continuación, cálculo las 4 raíces
        r1 = teorema_chino_restos(rp, rq, p, q) 
        r2 = teorema_chino_restos(p-rp, rq, p, q)
        r3 = (p*q - r1)
        r4 = (p*q - r2)
        
        return ([r1, r2, r3, r4]) 
    
    # Sino devuelve nada, no tiene raíces

In [27]:
raices(1,3,5)

[1, 11, 14, 4]

### Interpretación del Resultado Obtenido    

Quiero calcular $\sqrt{1}$ en $\mathbb{Z}_3$, en $\mathbb{Z}_5$ y en $\mathbb{Z}_{15}$

---

# Ejercicio 7

## Ejercicio 7.1

Implementa el Método de Fermat para factorización de enteros.

Ahora, pasamos a calcular el residuo cuadrático. Para ello, nos basamos en el **pdf de Jesús Miranda de Criptografía**, en concreto en la _página 198_ o _204_, abierto desde el visualizador de pdf's. 

### Pseudocódigo

1. Calculamos la raíz cuadrada de $n$, y llamamos x al entero inmediatamente superior.
2. Mientras $x²−n$ no sea un cuadrado perfecto, incrementamos $x$ en una unidad.
3. Cuando sea un cuadrado perfecto, le damos a $y$ el valor igual a la raíz cuadrada de $x²−n$
4. Con los valores de $x$ e $y$ conseguimos una factorización de $n$.

In [28]:
# -*- coding: utf-8 -*-

# Algoritmo que calcula la factorización de fermar de un número
def fermat(n):
    # Raiz cuadrada de n
    x = isqrt(n)+1
    y = 2
    
    # Hacemos uso de la función creada anteriormente para sacar raíces
    while y != isqrt(y)**2:
        y = x*x - n
        x = x+1
        
    return(x-1,isqrt(y))

In [29]:
fermat(6352351)

(2524, 135)

### Interpretación del resultado obtenido

Sea $n = 6352351$. Vamos a factorizarlo usando este método.
Se tiene que $\sqrt n = 2520.387$. Por tanto, le damos a $x$ el valor $2521$. Nos quedaría:

$x$ | $x²$-$n$ | ¿Es $x²-n$ cuadrado perfecto?
:---: | :---: | :---: 
2521 | 3090 | No 
2522 | 8133 | No
2523 | 13171 | No
2524 | 18225 | Si

In [30]:
# Para comprobar que es un cuadrado perfecto
import math
math.sqrt(18225)

135.0

## Ejercicio 7.2

Implementa el algoritmo de factorización Ω de Pollard.

Para ello nos basamos, en la referencia aportada por el _pdf's_. En concreto, nos basaremos en el [Algoritmo 3.9](http://cacr.uwaterloo.ca/hac/about/chap3.pdf)
 
### Pseudocódigo

**ENTRADA**: Un entero $n$.  
**SALIDA**: Un factor no trivial de $n$.  

1. Inicializar $a←2$, $b←2$.
2. Para $i = 1,2,...$ hacer:  
    2.1\. Inicializar $a←a^{2}+1 \pmod n$, $b←b^{2}+1 \pmod n$, $b←b^{2}+1 \pmod n$  
    2.2\. Obtener $d = mcd(a − b, n)$.  
    2.3\. Si $1 < d < n$ devolver $d$ y terminar.  
    2.4\. Si $d = n$ terminar la ejecución del algoritmo con un error.

In [31]:
# -*- coding: utf-8 -*-

# Calculamos el mcd(a,b)
def mcd (a, b):
    if a % b == 0:
        return b
    
    return mcd(b, a%b)

# Algoritmo que calcula el método de pollard
def alg_pollard(n):
    a = 2
    b = 2
    
    while (True) :
        a = ((a**2)+1) % n
        b = ((b**2)+1) % n
        b = ((b**2)+1) % n
        
        # Obtener el mcd(a-b, n)
        d = mcd(a-b, n)
        
        if (1<d and d<n):
            return d
        
        # Terminar la ejecución del algoritmo con un error
        if (d == n):
            print("Error")
            return None

In [32]:
alg_pollard(455459)

743

### Interpretación del resultado obtenido

A continuación, vamos a comprobar de donde ha salido el resultado, recordar que estamos en $\pmod 455459$:

a | b | d
:---: | :---: | :---: | 
5 = $(2^{2}+1)$ | 26 | 1 = $mcd(-21, 455459)$
26 = $(5^{2}+1)$ | 2871 | 1 = $mcd(-2845, 455459)$
677 = $(26^{2}+1)$ | 179685 | 1 = $mcd(-179008, 455459)$
2871 = $(677^{2}+1)$| 155260 | 1 = $mcd(-152389, 455459)$
44380 = $(2871^{2}+1)$ | 416250 |1 = $mcd(-371870, 455459)$
179685 = $(44380^{2}+1)$ | 43670 | 1 = $mcd(136015, 455459)$
121634 = $(179685^{2}+1)$ | 164403 | 1 = $mcd(-42769, 455459)$
155260 = $(121634^{2}+1)$ | 247944 | 1 = $mcd(-92684, 455459)$
44567 = $(15526^{2}+1)$ | 68343 | 743 = $mcd(-23776, 455459)$

Para comprobar que el resultado es correcto, lo comprobamos haciendo en **GAP**:
> FactorsRho(n);

---

# Ejercicio 8

Compara los tiempos de ejecución de tus implementaciones con las de tus compañeros y con las
primitivas de algunos paquetes de cálculo simbólico como (GAP, maxima...)

Primero, calcularemos los tiempos que tardan mis algoritmos.

In [33]:
import time

tiempos = []

# Algoritmo extendido de euclides
t1 = time.time()
alg_euclides(28564333765949, 123456789101119)
t2 = time.time()
tiempos.append("Euclides: " + str(t2 - t1))

# Algoritmo Inverso
t1 = time.time()
alg_inverso(28564333765949, 123456789101119)
t2 = time.time()
tiempos.append("Algoritmo Inverso: " + str(t2 - t1))

# Algoritmo Potencia
t1 = time.time()
alg_potencia(8590365927553, 28564333765949, 123456789101119)
t2 = time.time()
tiempos.append("Algoritmo Potencia: " + str(t2 - t1))

# Miller-rabin
t1 = time.time()
alg_miller_rabin1(123456789101119, 10)
t2 = time.time()
tiempos.append("Algoritmo Miller-Rabin: " + str(t2 - t1))

# Paso enano paso gigante
t1 = time.time()
paso_enano_paso_gigante(859036592, 285643337, 46381)
t2 = time.time()
tiempos.append("Paso enano paso gigante: " + str(t2 - t1))

# Residuo Cuadratico
t1 = time.time()
residuo_cuadratico(123458977,8590365927553)
t2 = time.time()
tiempos.append("Residuo cuadrático: " + str(t2 - t1))

# Raíces
t1 = time.time()
raices(123458977,8590365927553, 28564333765949)
t2 = time.time()
tiempos.append("Raices: " + str(t2 - t1))

# Fermat
t1 = time.time()
fermat(1234568977)
t2 = time.time()
tiempos.append("Fermat: " + str(t2 - t1))

# Pollard
t1 = time.time()
alg_pollard(8590365927554)
t2 = time.time()
tiempos.append("Pollard: " + str(t2 - t1))

print(tiempos)

['Euclides: 0.00011491775512695312', 'Algoritmo Inverso: 9.250640869140625e-05', 'Algoritmo Potencia: 0.00014209747314453125', 'Algoritmo Miller-Rabin: 0.0007910728454589844', 'Paso enano paso gigante: 0.006052494049072266', 'Residuo cuadrático: 0.0003809928894042969', 'Raices: 0.0006351470947265625', 'Fermat: 2.684865951538086', 'Pollard: 4.76837158203125e-05']


En el enunciado, se nos pide que compararemos con las primitivias de algunos paquetes de cálcula simbólico como $GAP$ o $Maxima$. Pues bien, al haber realizado los ejercicios anteriores en $GAP$, el tiempo que tarda en ejecutarlos, es alrededor de cero. Por tanto, he considerado despreciar tal comparación y centrarme en la de mis compañeros. Del mismo modo, probé a realizarlos en $Maxima$.

Para realizar la comparación con mis compañeros, se han usado los mismos números. Lo único que difieren son los algoritmos y la máquina donde se han sacado los tiempos:

**Algoritmo** | **Compañero 1** | **Compañero 2** | **Yo**
:---: | :---: | :---: 
Euclides | 7.46e-05 | 0.00053 | 4.91e-05
Inverso | 7.96e-05 | 1.41E-05 | 4.34e-05
Potencia | 0.00011 | 7.20E-05 | 7.25e-05
Miller-Rabin | 0.000603 | 0.000622 | 0.000391
Paso Enano Paso Gigante | 0.000329 | 0.00154 | 0.0036
Residuo Cuadratico (raíz de $p$) | 9.49e-05 | 2.19E-05 | 6.19e-05 
Raices Modulares (raíz de $pq$) | 0.000143 | 2.72E-05 | 7.41e-05
Fermat | 3.91 | 1.710 | 2.63
Pollard | 9.87e-05 | 2.00E-05 |  5.32e-05
**Media** | **4.35e-01** | **1.90e-01** | **2.93e-01**

Para el **Compañero 1**, como se puede apreciar, en casi todos los algoritmos mi ordenador obtiene un mejor tiempo. Esto es debido, a que las prestaciones ofrecidas por mi ordenador son mejores incluso usando una máquina virtual, y también que mi ordenador tiene 2-3 años y el de mi compañero, 7. Por ejemplo, en el algoritmo de **paso enano paso gigante**, obtengo un peor resultado, esto es así ya que mi algoritmos tiene orden de eficiencia alrededor del cuadrático, mientras que mi compañero tiene un algoritmo con un orden de eficiencia menor.

Pero como se la **Compañera 2**, obtiene los mejores resultados, obtiendo $0.1s$ menos que mis algoritmos.

Además, junto a mi **Compañero 1**, hemos realizado otro tipo de comparación. Básicamente consiste, en que tanto él como yo, hemos ejecutado los algoritmos de ambos. 

_**Nota**: PC.1 será el PC del Compañero 1 y PC.YO será mi ordenador._

**Algoritmo** | **Yo en PC.1** | **Yo PC.YO** | **Compañero 1 en PC.1** | **Compañero 1 en PC.YO**
:---: | :---: | :---: | :---: 
Euclides | 0.000281| 4.91E-05 | 8.44E-05 | 4.12E-05
Inverso | 0.00028 | 4.34E-05 | 7.70E-05 | 4.43E-05
Potencia | 0.00041 | 7.25E-05 | 0.0001084 | 6.56E-05
Miller-Rabin | 0.0014 | 0.000391 | 0.000554561615 | 0.0003833
Paso Enano Paso Gigante | 0.0133 | 0.003590345383 | 0.0003039 | 0.000207
Residuo Cuadratico (raíz de $p$) | 0.000307 | 6.20E-05 | 8.54E-05 | 5.46E-05
Raices Modulares (raíz de $pq$) | 0.000340 | 7.41E-05 | 0.000125 | 9.01E-05
Fermat | 3.896 | 2.633 | 3.8872 | 2.787
Pollard | 7.94E-05 | 5.32E-05 |  0.000184 | 4.39E-05
**Media** | **4.34e-01** | **2.93E-01** | **4.32E-01** | **3.10E-01**

Una vez obtenida la tabla vamos a dar una conclusión. Como se puede ver, los algoritmos tanto míos como de mi compañero en la máquina de mi compañero, tardan aproximadamente lo mismo. Igual ocurre en mi ordenador. Sin embargo, cuando lanzamos un mismo algoritmo en distintas máquinas, mi ordenador tarda menos que en el de mi compañero. Esto quiere decir que aunque tengamos algoritmos parecidos, depende de la carga y la máquina en que se este ejecutando, y como dije antes, mi ordenador tiene mejores prestaciones que el de mi compañero. 